File: internal/mdx/images.go
// Package mdx scans and rewrites image references in Markdown source.
//
// v0.1 handles the two common forms found in AI-generated Markdown:
//
//	![alt](url)            // CommonMark inline image
//	![alt](url "title")    // with title
//	<img src="url" .../>   // raw HTML
//
// Reference-style images (![alt][ref] + [ref]: url) are out of scope for v0.1.
package mdx

import (
	"regexp"
	"strings"
)

// ImageRef describes one image reference inside Markdown source.
type ImageRef struct {
	URL      string // raw destination as written
	Alt      string // alt text (may be empty)
	Title    string // optional title
	Start    int    // byte offset of the URL within the source
	End      int    // byte offset after the URL
	IsHTML   bool   // true when matched from <img src="...">
}

// inlineImage matches ![alt](url) and ![alt](url "title").
// Group layout: 1=alt, 2=url, 3=optional title (without quotes).
var inlineImage = regexp.MustCompile(
	`!\[([^\]]*)\]\(\s*([^)\s]+)(?:\s+"([^"]*)")?\s*\)`,
)

// htmlImage matches <img ... src="url" ... />. Group 1 is the URL.
var htmlImage = regexp.MustCompile(`(?i)<img\b[^>]*?\bsrc\s*=\s*["']([^"']+)["'][^>]*>`)

// Scan returns every image reference in src.
func Scan(src []byte) []ImageRef {
	var refs []ImageRef
	for _, m := range inlineImage.FindAllSubmatchIndex(src, -1) {
		ref := ImageRef{
			Alt:   string(src[m[2]:m[3]]),
			URL:   string(src[m[4]:m[5]]),
			Start: m[4],
			End:   m[5],
		}
		if m[6] != -1 {
			ref.Title = string(src[m[6]:m[7]])
		}
		refs = append(refs, ref)
	}
	for _, m := range htmlImage.FindAllSubmatchIndex(src, -1) {
		refs = append(refs, ImageRef{
			URL:    string(src[m[2]:m[3]]),
			Start:  m[2],
			End:    m[3],
			IsHTML: true,
		})
	}
	return refs
}

// Rewrite replaces each image URL using mapper. If mapper returns the
// original URL unchanged, the source byte range is left intact. URLs that
// mapper returns as "" are also left unchanged.
func Rewrite(src []byte, mapper func(ref ImageRef) string) []byte {
	refs := Scan(src)
	if len(refs) == 0 {
		return src
	}
	// Apply replacements right-to-left so earlier offsets stay valid.
	sorted := make([]ImageRef, len(refs))
	copy(sorted, refs)
	for i := 0; i < len(sorted); i++ {
		for j := i + 1; j < len(sorted); j++ {
			if sorted[j].Start > sorted[i].Start {
				sorted[i], sorted[j] = sorted[j], sorted[i]
			}
		}
	}
	out := make([]byte, len(src))
	copy(out, src)
	for _, ref := range sorted {
		newURL := mapper(ref)
		if newURL == "" || newURL == ref.URL {
			continue
		}
		out = append(out[:ref.Start], append([]byte(newURL), out[ref.End:]...)...)
	}
	return out
}

// IsRemote reports whether url is an http(s) or data URL.
func IsRemote(url string) bool {
	u := strings.ToLower(url)
	return strings.HasPrefix(u, "http://") ||
		strings.HasPrefix(u, "https://") ||
		strings.HasPrefix(u, "data:")
}

// AssetURIPrefix is the canonical scheme for AIMD-internal assets.
const AssetURIPrefix = "asset://"

// IsAssetURI reports whether url is an asset:// reference.
func IsAssetURI(url string) bool { return strings.HasPrefix(url, AssetURIPrefix) }

// AssetURIID extracts the id portion of an asset:// URL, or "".
func AssetURIID(url string) string {
	if !IsAssetURI(url) {
		return ""
	}
	id := strings.TrimPrefix(url, AssetURIPrefix)
	// Strip query/fragment if any.
	if i := strings.IndexAny(id, "?#"); i >= 0 {
		id = id[:i]
	}
	return id
}

File: crates/aimd-mdx/src/images.rs
use regex::Regex;
use std::cmp::Reverse;
use std::sync::OnceLock;

/// Describes one image reference inside Markdown source.
#[derive(Debug, Clone)]
pub struct ImageRef {
    pub url: String,
    pub alt: String,
    pub title: String,
    pub start: usize,
    pub end: usize,
    pub is_html: bool,
}

fn inline_image_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r#"!\[([^\]]*)\]\(\s*([^)\s]+)(?:\s+"([^"]*)")?\s*\)"#).unwrap())
}

fn html_image_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r#"(?i)<img\b[^>]*?\bsrc\s*=\s*["']([^"']+)["'][^>]*>"#).unwrap())
}

/// Returns every image reference in src.
pub fn scan(src: &[u8]) -> Vec<ImageRef> {
    let text = match std::str::from_utf8(src) {
        Ok(s) => s,
        Err(_) => return Vec::new(),
    };
    let mut refs = Vec::new();

    for cap in inline_image_re().captures_iter(text) {
        let m = cap.get(0).unwrap();
        let url_m = cap.get(2).unwrap();
        refs.push(ImageRef {
            alt: cap
                .get(1)
                .map(|m| m.as_str().to_string())
                .unwrap_or_default(),
            url: url_m.as_str().to_string(),
            title: cap
                .get(3)
                .map(|m| m.as_str().to_string())
                .unwrap_or_default(),
            start: url_m.start(),
            end: url_m.end(),
            is_html: false,
        });
        let _ = m;
    }

    for cap in html_image_re().captures_iter(text) {
        let url_m = cap.get(1).unwrap();
        refs.push(ImageRef {
            url: url_m.as_str().to_string(),
            alt: String::new(),
            title: String::new(),
            start: url_m.start(),
            end: url_m.end(),
            is_html: true,
        });
    }

    refs
}

/// Replaces each image URL using mapper.
/// If mapper returns "" or the original URL unchanged, the source byte range is left intact.
/// Applies replacements right-to-left so earlier offsets stay valid.
pub fn rewrite(src: &[u8], mut mapper: impl FnMut(&ImageRef) -> String) -> Vec<u8> {
    let refs = scan(src);
    if refs.is_empty() {
        return src.to_vec();
    }

    let text = std::str::from_utf8(src).unwrap_or("");

    // Sort descending by start so we apply replacements right-to-left.
    let mut sorted = refs;
    sorted.sort_by_key(|r| Reverse(r.start));

    let mut out = text.to_string();
    for ref_item in &sorted {
        let new_url = mapper(ref_item);
        if new_url.is_empty() || new_url == ref_item.url {
            continue;
        }
        out.replace_range(ref_item.start..ref_item.end, &new_url);
    }
    out.into_bytes()
}

/// Reports whether url is an http(s) or data URL.
pub fn is_remote(url: &str) -> bool {
    let u = url.to_lowercase();
    u.starts_with("http://") || u.starts_with("https://") || u.starts_with("data:")
}

/// Reports whether url is an asset:// reference.
pub fn is_asset_uri(url: &str) -> bool {
    url.starts_with("asset://")
}

/// Extracts the id portion of an asset:// URL, or "".
pub fn asset_uri_id(url: &str) -> &str {
    if !is_asset_uri(url) {
        return "";
    }
    let id = &url["asset://".len()..];
    // Strip query/fragment if any.
    if let Some(i) = id.find(['?', '#']) {
        &id[..i]
    } else {
        id
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scan_inline() {
        let src =
            b"# Title\n\n![cover](./cover.png)\n\nSome text ![diag](images/d.svg \"hello\") end.\n";
        let refs = scan(src);
        assert_eq!(refs.len(), 2, "want 2 refs, got {}", refs.len());
        assert_eq!(refs[0].url, "./cover.png");
        assert_eq!(refs[0].alt, "cover");
        assert_eq!(refs[1].url, "images/d.svg");
        assert_eq!(refs[1].title, "hello");
    }

    #[test]
    fn test_scan_html() {
        let src = b"<img src=\"a.png\" alt=\"x\" />";
        let refs = scan(src);
        assert_eq!(refs.len(), 1);
        assert_eq!(refs[0].url, "a.png");
        assert!(refs[0].is_html);
    }

    #[test]
    fn test_rewrite() {
        let src = b"![a](one.png) and ![b](two.png)";
        let got = rewrite(src, |r| format!("asset://{}", r.url));
        assert_eq!(
            std::str::from_utf8(&got).unwrap(),
            "![a](asset://one.png) and ![b](asset://two.png)"
        );
    }

    #[test]
    fn test_rewrite_no_change() {
        let src = b"![a](http://example.com/x.png)";
        let got = rewrite(src, |r| {
            if is_remote(&r.url) {
                String::new()
            } else {
                format!("asset://{}", r.url)
            }
        });
        assert_eq!(got, src.to_vec(), "remote URL should be untouched");
    }

    #[test]
    fn test_asset_uri() {
        assert!(is_asset_uri("asset://abc"));
        assert_eq!(asset_uri_id("asset://abc?x=1"), "abc");
    }
}
